I2C Implementation

Disteroas opened the pull request:
Abbiamo creato motorI2C.cpp e motorI2C.h, ma in modo indipendente dai file già presenti nella cartella (ad esempio motor.cpp o motor.h)

Disteroas committed:
Creazione file motorI2C.cpp e .h

luca-byte commented:
Adesso (nei prossimi giorni) guardo con più attenzione ma mi sembra un buon inizio. Ottimo lavoro! Per la gestione dei motori tanto si eseguiranno le chiamate alla libreria SmartMotor quindi direi che i file di cui parlavate si possono al momento ignorare

Disteroas commented:
> Adesso (nei prossimi giorni) guardo con più attenzione ma mi sembra un buon inizio. Ottimo lavoro! Per la gestione dei motori tanto si eseguiranno le chiamate alla libreria SmartMotor quindi direi che i file di cui parlavate si possono al momento ignorare

Si devo dire che chatgpt si è impegnata molto...

mar-ceve commented:
Sotto consiglio di @luca-byte, ho fatto alcune modifiche al codice:

- Ho aggiunto un controllo nel loop per fermare i motori se non arrivano nuovi comandi entro un certo tempo.
- Ho cambiato il modo in cui si gestiva la velocità dei motori, usando direttamente i metodi della libreria SmartMotor invece di salvare i valori in variabili.
- Ho sistemato il codice per far sì che i motori vengano aggiornati subito quando ci sono nuovi dati, usando un flag per indicare quando arrivano comandi nuovi.

Scusate sono poco pratico con github :/, come vi condivido il file??

Disteroas commented:
> Sotto consiglio di @luca-byte, ho fatto alcune modifiche al codice:
> 
> * Ho aggiunto un controllo nel loop per fermare i motori se non arrivano nuovi comandi entro un certo tempo.
> * Ho cambiato il modo in cui si gestiva la velocità dei motori, usando direttamente i metodi della libreria SmartMotor invece di salvare i valori in variabili.
> * Ho sistemato il codice per far sì che i motori vengano aggiornati subito quando ci sono nuovi dati, usando un flag per indicare quando arrivano comandi nuovi.
> 
> Scusate sono poco pratico con github :/, come vi condivido il file??

#toro

luca-byte commented:
> Sotto consiglio di @luca-byte, ho fatto alcune modifiche al codice:
> 
> - Ho aggiunto un controllo nel loop per fermare i motori se non arrivano nuovi comandi entro un certo tempo.
> - Ho cambiato il modo in cui si gestiva la velocità dei motori, usando direttamente i metodi della libreria SmartMotor invece di salvare i valori in variabili.
> - Ho sistemato il codice per far sì che i motori vengano aggiornati subito quando ci sono nuovi dati, usando un flag per indicare quando arrivano comandi nuovi.
> 
> Scusate sono poco pratico con github :/, come vi condivido il file??

Forse non mi è chiara la domanda. Direi che modifichi il file su questo branch e poi effettui un commit. Trovi una guida a riguardo su Outline sezione Software Guides

Disteroas committed:
Prime modifiche ai file motorI2C

marc0777 commented on MotorI2C/motorI2C.h:
```diff
@@ -0,0 +1,25 @@
+    void loopUpdate();                    // Regular update function to be called in the main loop
+
+  private:
+    int _slaveAddress;
+    SmartMotor* _leftMotor;               // Pointer to left motor
+    SmartMotor* _rightMotor;              // Pointer to right motor
+    bool _newDataAvailable;               // Flag to indicate new data has been received
+    unsigned long _lastCommandTime;       // Time of the last received command
```
It's bad practice in C++ to start variables with an underscore, unless you have good reasons to do so (developing standard libraries, OS/compiler components, ...).

If you want to clearly distinguish member variables from the ones defined in the function you could prefix them with `m_` (instead of just `_`) which is the most used practice in C++. However this wasn't done anywhere else in the codebase and isn't common in the Arduino space.

marc0777 commented on MotorI2C/motorI2C.cpp:
```diff
@@ -0,0 +1,61 @@
+void MotorI2C::setupI2C(SmartMotor* leftMotor, SmartMotor* rightMotor) {
+  _leftMotor = leftMotor;
+  _rightMotor = rightMotor;
+  Wire.begin(_slaveAddress);  // Initialize the I2C as a slave
+  Wire.onReceive(receiveData);  // Set the receive function to handle incoming data
+  Wire.onRequest(sendTelemetry); // Set the request function to handle telemetry data
```
Be very careful with this! Initialising the `Wire` interface inside an "hidden" function risks breakage that's very difficult to debug.

It's always best to initialise every hardware interface in the Arduino `setup()` function, to absolutely avoid conflicts that would lead the MCU to crash without giving no error on the console.

Also, it's a good practice to pass the `Wire` interface from the outside, by accepting a `TwoWire`. This way you can change the physical I2C interface configuration without having to change this class in any way!

luca-byte replied:
Reading this I am not getting why would it be worse to call this function in the setup() instead of initializing the Wire directly there, other than the fact that we would need to change the class as you mentioned in the last paragraph.

marc0777 replied:
In this case, it's mainly to avoid user error. Wire functions must be called in the correct moment and in the correct order, and any mistake often leads to unexpected panics (crash with no logs). I'll list two of the possibly many scenarios that could happen.

### Multiple `begin`
Let's say that you have configured the `MotorI2C` class, and in Arduino's `setup()` you call `setupI2C()`. Then after a while you realise you also want to use a I2C sensor, so you import its library and start using. You read in it's docs that you should call `Wire.begin()` without params to initialise the com, so you execute it immediately after calling `setupI2C()`. This will lead to a panic with no logs! You can mitigate this passing the `TwoWire` interface and make sure to use two different physical interfaces, but it's much harder having to remember if the `begin()` is hidden inside another class.

### Changing pins
You need to free the pins currently used by the I2C to connect another device that could only work with those pins. So you [read the docs](https://arduino-pico.readthedocs.io/en/latest/wire.html) (let's say for the Pi Pico) and see that you should call `Wire.setSDA()` and `Wire.setSCL()` immediately before the call to `Wire.begin()` however you don't know exactly where that is called since it's hidden inside another class, and you end up putting them after `setupI2C()`. You start your program and... panic!

Even worse, if you look for the ESP32 ([docs here](https://arduino-pico.readthedocs.io/en/latest/wire.html)), you could find that to change the pins you should change the call to `Wire.begin()` and add the pins there. So you'd end up even more domain violations!

marc0777 commented on MotorI2C/motorI2C.cpp:
```diff
@@ -0,0 +1,61 @@
+#include "MotorI2C.h"
+#include "SmartMotor.h"
+
+MotorI2C::MotorI2C(int slaveAddress) {
+  _slaveAddress = slaveAddress;
+  _leftMotor = nullptr;  // Pointer to SmartMotor instance for left motor
+  _rightMotor = nullptr; // Pointer to SmartMotor instance for right motor
+  _newDataAvailable = false;
+  _lastCommandTime = 0;
+}
```
If you end up removing the `_` from the member variables, remember to update this, which would become something like this:
```suggestion
MotorI2C::MotorI2C(int slaveAddress) {
  this->slaveAddress = slaveAddress;
  leftMotor = nullptr;  // Pointer to SmartMotor instance for left motor
  rightMotor = nullptr; // Pointer to SmartMotor instance for right motor
  newDataAvailable = false;
  lastCommandTime = 0;
}
```

Another way to implement the same constructor in a more compact way could also be:
```suggestion
MotorI2C::MotorI2C(int slaveAddress)
    : slaveAddress(slaveAddress),
      newDataAvailable(false),
      lastCommandTime(0)
{
}
```

Notice that you don't need to set the `nullptr`, since you'll be overwriting those variables anyway before using!

marc0777 commented on MotorI2C/motorI2C.cpp:
```diff
@@ -0,0 +1,61 @@
+  _lastCommandTime = 0;
+}
+
+// Setup I2C communication and initialize motors
```
All member functions should ideally be preceded by a Doxygen compliant comment, which documents not only the function itself but also its parameters and return. For example, for this function it could look like this:

```suggestion
/**
 * Setup I2C communication and initialise motors.
 * @param leftMotor instance of SmartMotor to be used as left motor. 
 * @param leftMotor instance of SmartMotor to be used as right motor. 
 */
```

marc0777 reviewed (commented):
Hello! Sorry it took so long :)

I've added a few comments of things that in my opinion can be improved. Overall I'd say it's a great start, and the content of the comments are mainly code quality improvements, and some tips to avoid future headaches (after having spent a lot of time on similar issues myself).

I'm writing in English since the codebase is public and not everyone in the team always was Italian mother tongue, and also makes it more appealing here in GitHub. However this should be discussed/agreed with current team members!

luca-byte commented on MotorI2C/motorI2C.cpp:
```diff
@@ -0,0 +1,61 @@
+// Function to be called in the main loop for regular updates
+void MotorI2C::loopUpdate() {
+  // If new data has been received, update the motor control
+  if (_newDataAvailable) {
+    _leftMotor->update();
+    _rightMotor->update();
+    _newDataAvailable = false;
+  }
```
Be careful, this is not what the code in PicoLowLevel is doing with the motors [see here](https://github.com/Team-Isaac-Polito/PicoLowLevel/blob/bfe21c8aba421f755014b434a067ea342692a642/PicoLowLevel.ino#L193-L194)

However I would suggest against coupling together the SmartMotor "library" and the I2C communication. 

In my opinion the following would be clearer and simpler to manage:
- the I2C receives asyncrhonously the data on the I2C and raises a flag (`newDataAvailable`) 
- In the `loop()` a function is called to check if data is available (e.g. `bool isAvailablke(float* left, float* right)`) 
- If there is data we send it to the motor
- Every loop iteration we update the motor
- We implement the safety check that stops the motor in the loop

I am dubious about your approach because it links two different functionalities that can be separated. In my opinion it would be better, if we want to do something similar, to include a function like this in the SmartMotor class

luca-byte reviewed (commented):
I left a comment about the approach to take in the library. Tell me what do you think.

Disteroas commented:
Sorry for getting to your suggestions just now. Tomorrow, Marcello and I will try to fix all (hopefully!) the issues

Disteroas committed:
Oxygen compliant comment MotorI2C/motorI2C.cpp

Co-authored-by: Marco Fincato <marcofincato777@gmail.com>

Disteroas committed:
Files ready for first tests

We've modified motorI2C files following all the suggestions (really appreciated) and we've created 2 files: testESP.ino and testPICO.ino for future tests.

Disteroas committed:
Update motorI2C.cpp

Disteroas committed:
Update motorI2C.h

Disteroas commented:
We tried to run a test on an ESP and a Pico, adding the libraries that we've found on gh (in PLL), but there were a lot of troubles. (and we don't know why)

luca-byte commented:
> We tried to run a test on an ESP and a Pico, adding the libraries that we've found on gh (in PLL), but there were a lot of errors. (and we don't know why)

It is really difficult to help you with so little info about the errors. Would you mind giving us more info?

I will take a look to your modifications

lorixhack reviewed (commented):
It seems pretty good to me!

luca-byte commented on MotorI2C/motorI2C.cpp:
```diff
@@ -0,0 +1,57 @@
+}
+
+// Check if new data is available and return the motor speeds
+bool MotorI2C::isAvailable(float* leftSpeed, float* rightSpeed) {
+  if (newDataAvailable) {
+    *leftSpeed = receivedLeftSpeed;
+    *rightSpeed = receivedRightSpeed;
+    newDataAvailable = false;
+    return true;
+  }
+  return false;
+}
```
We can delete this function if we merge all the update functionality in the same function called in the loop.

luca-byte commented on MotorI2C/motorI2C.cpp:
```diff
@@ -0,0 +1,57 @@
+
+// Function to send telemetry (current speeds) back to the master
+void MotorI2C::sendTelemetry() {
+  float leftSpeed = receivedLeftSpeed;
+  float rightSpeed = receivedRightSpeed;
```
The telemetry should use data obtained from the motor, to detect any discrepancy between the commands and the effective speed of the motor using `SmartMotor#getSpeed()`

Disteroas replied:
We've commented receiveData and we updated sendTelemetry using SmartMotor pointers

luca-byte commented on MotorI2C/motorI2C.cpp:
```diff
@@ -0,0 +1,57 @@
+
+// Setup I2C communication
+void MotorI2C::begin() {
+  wireInstance.begin(slaveAddress);  // Initialize the I2C as a slave
```
The wire is already started in the `setup()`. This is the error Marco was talking about

luca-byte commented on MotorI2C/motorI2C.cpp:
```diff
@@ -0,0 +1,57 @@
+void MotorI2C::begin() {
+  wireInstance.begin(slaveAddress);  // Initialize the I2C as a slave
+  wireInstance.onReceive(receiveData);  // Set the receive function to handle incoming data
+  wireInstance.onRequest(sendTelemetry); // Set the request function to handle telemetry data
```
Since we have the motors here we can decide if we want to `SmartMotor#begin()` and `SmartMotor#calibrate()` here or outside the library, but at the moment these necessary functions are not called anywhere

Disteroas replied:
We are trying to fix all problems, in this case we understood what is the issue, but we don't know which solution is better. What do you suggest?

mar-ceve replied:
You said that we have to decide where to call this functions (in the library or outside) but we don't know which option is better. We were asking you for a suggestion.

luca-byte replied:
Ops, I read a not which was not there 😅

luca-byte reviewed (changes requested):
First and foremost, great job!! 💯 

I decided to leave some comments about a different decision in the final implementation as discussed in today's meeting. I tried to be exhaustive in enumerating all the changes, but verify that all the complementary changes are also made. 

Remember also to **include the SmartMotor** inside the project or otherwise the library cannot work. You can take it from the relative test folder in this repository, remember to remove the `.ino` file and the `Sensor.*` files

luca-byte commented:
That is, unless @marc0777, you have a better idea to deal with the async request of telemetry

Disteroas committed:
Update MotorI2C/motorI2C.cpp

Co-authored-by: Luca De Paola <60526200+luca-byte@users.noreply.github.com>

Disteroas committed:
Update motorI2C.cpp

We've corrected some of the issues commented by @luca-byte. In the next days we'll correct all of them

Disteroas committed:
Update motorI2C.h

Pitosoro committed:
updated motorI2C

Disteroas committed:
Update motorI2C.cpp

Disteroas committed:
Update motorI2C.h

Disteroas committed:
Delete motorI2C.h

There were 2 copies of the file, I've deleted one of them

Disteroas committed:
Delete motorI2C.cpp

There were 2 copies of the file, I've deleted one of them

Disteroas committed:
Delete MotorI2C/testESP.ino

We've created a new branch on mainboard for I2C and pasted this file there

luca-byte committed:
Merge remote-tracking branch 'origin/main' into motorsboard

luca-byte committed:
feat: Finalize I2C library
